third_party/move/move-vm/metrics/src/lib.rs

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use once_cell::sync::Lazy;
+use prometheus::{register_histogram_vec, HistogramTimer, HistogramVec};
+
+/// Helper trait to encapsulate [HistogramVec] functionality. Users can use this trait to time
+/// different VM parts collecting metrics for different labels. Use wisely as timers do introduce
+/// an overhead, so using on extremely hot path is not recommended.
```

Comment:
I think that this is too optimistic of a wording, let's at least remove word "extremely".  Even a moderate hot paths can cause slowdown, as histogram is not that cheap. 

Things in this PR are probably fine (I assume type_to_type_tag is not called too often?), but it is easy do add overhead (i.e
 Cache hot for data would probably be questionable)

Reply:
Good point! 

Was testing with e2e, data accesses are indeed something we do not want to record, as there is a bit of regression. At least recording gas traversal & loading/verification miss + total execution time should be enough